web/api/tests/web_api.c

```diff
@@ -126,43 +214,239 @@ static void build_request(struct web_buffer *wb, const char *url, bool use_cr, s
     return w;
 }
 
-static void post_test_cleanup(struct web_client *w)
+static void destroy_web_client(struct web_client *w)
 {
     buffer_free(w->response.data);
     buffer_free(w->response.header);
     buffer_free(w->response.header_output);
     free(w);
-    free(localhost);
 }
 
+// ---------------------------------- Parameterized test-families -----------------------------------------------------
+// There is no way to pass a parameter block into the setup fixture, we would have to patch CMocka and maintain it
+// locally. (The void **current_state in _run_group_tests would be set from a parameter). This is unfortunate as a
+// parameteric unit-tester needs to be to pass parameters to the fixtures. We are faking this by calculating the
+// space of tests in the launcher, passing an array of identical unit-tests to CMocka and then counting through the
+// parameters in the shared state passed between tests. To initialise this counter structure we use this global to
+// pass from the launcher (test-builder) to the setup-fixture.
+
+void *shared_test_state = NULL;
+
+// -------------------------------- Test family for /api/v1/info ------------------------------------------------------
+
+struct test_def {
+    size_t num_headers; // Index coordinate
+    size_t prefix_len;  // Index coordinate
+    char name[80];
+    size_t full_len;
+    struct web_client *instance; // Used within this single test
+    bool completed;
+    struct test_def *next, *prev;
+};
+
 static void api_info(void **state)
 {
+    (void)state;
+    struct test_def *def = (struct test_def *)shared_test_state;
+    shared_test_state = def->next;
+
+    if (def->prev != NULL && !def->prev->completed && strlen(log_buffer) > 0) {
+        printf("Log of failing case %s:\n", def->prev->name);
+        puts(log_buffer);
+    }
+    log_buffer[0] = 0;
+    if (localhost != NULL)
+        free(localhost);
+    localhost = calloc(1,sizeof(RRDHOST));
+
+    def->instance = setup_fresh_web_client();
+    build_request(def->instance->response.data, "/api/v1/info", true, def->num_headers);
+    def->instance->response.data->len = def->prefix_len;
+
+    info("Buffer contains: %s [first %zu]", def->instance->response.data->buffer, def->prefix_len);
+    if (def->prefix_len == def->full_len) {
+        expect_value(__wrap_web_client_api_request_v1, host, localhost);
+        expect_value(__wrap_web_client_api_request_v1, w, def->instance);
+        expect_string(__wrap_web_client_api_request_v1, url, "info");
+    }
+
+    web_client_process_request(def->instance);
+
+    if (def->prefix_len == def->full_len)
+        assert_int_equal(def->instance->flags & WEB_CLIENT_FLAG_WAIT_RECEIVE, 0);
+    else
+        assert_int_equal(def->instance->flags & WEB_CLIENT_FLAG_WAIT_RECEIVE, WEB_CLIENT_FLAG_WAIT_RECEIVE);
+    assert_int_equal(def->instance->mode, WEB_CLIENT_MODE_NORMAL);
+    printf("decoded: %s\n", def->instance->decoded_query_string);
```

Comment:
In every test it prints an empty string. It just clutters the output. Do you really need it?

Reply:
Now it's much better. Passed tests are clean.